sherpa/fit.py

```diff
@@ -1119,29 +1235,63 @@ def fit(self, outfile=None, clobber=False, numcores=1):
         #
-        if not iterable(dep) or len(dep) == 0:
+        if not np.iterable(dep) or len(dep) == 0:
             raise FitErr('nobins')
 
-        if ((iterable(staterror) and 0.0 in staterror) and
+        if ((np.iterable(staterror) and 0.0 in staterror) and
                 isinstance(self.stat, Chi2) and
                 type(self.stat) != Chi2 and
                 type(self.stat) != Chi2ModVar):
             raise FitErr('binhas0')
 
         init_stat = self.calc_stat()
-        # output = self.method.fit ...
-        tmp = self._iterfit._get_callback(outfile, clobber)
-        output = self._iterfit.fit(tmp,
-                                   self.model.thawedpars,
-                                   self.model.thawedparmins,
-                                   self.model.thawedparmaxes)
-        # LevMar always calculate chisquare, so call calc_stat
-        # just in case statistics is something other then chisquare
-        self.model.thawedpars = output[1]
-        tmp = list(output)
-        tmp[2] = self.calc_stat()
-        output = tuple(tmp)
-        # end of the gymnastics 'cause one cannot write to a tuple
+
+        # Using close_on_exit is not ideal but the output is optional,
+        # and it depends on what the input argument was, so it's
+        # awkward to do this with a context manager.
+        #
+        close_on_exit = True
+        if outfile is not None:
+
+            # If this is a "file handle" then skip the clobber check
+            # and mark as something we do not close on exit.
+            #
+            if isinstance(outfile, WriteableTextFile):
+                fh = outfile
+                close_on_exit = False
+            else:
+                # os.path.isfile and open accepts strings and Path objects
+                if not clobber and os.path.isfile(outfile):
+                    raise FitErr('noclobererr', str(outfile))
+
+                fh = open(outfile, mode='w', encoding="ascii")
+
+            # Write out the header line
+            names = ['#', 'nfev', 'statistic']
+            names.extend(par.fullname
+                         for par in self.model.get_thawed_pars())
+
+            fh.write(' '.join(names) + '\n')
+
+        else:
+            fh = None
+
+        cb = self._iterfit._get_callback(fh=fh)
+        output_orig = self._iterfit.fit(cb,
+                                        self.model.thawedpars,
+                                        self.model.thawedparmins,
+                                        self.model.thawedparmaxes)
+
+        (status, newpars, fval, msg, imap) = output_orig
+
+        # We can not just call cb(newpars) - which would have meant
+        # that writing out the final statistic below is not necessary
+        # - since the statistic used there does not necessarily match
+        # the user-requested statistic. We could change cb.stat but
+        # that seems more effort than it's worth.
+        #
```

Comment:
I don't know. I didn't write this code, I am just trying to add documentation to it. Maybe I'm reading the code wrong, but I am pretty sure that you can the statistic has somehow changed. It's something I want to talk to @anetasie about at some point.

Reply:
I will say that I am sure I saw, in testing this, cases where the stats did  differ - that is `fval` and `fval_new` were different. However, I can't seem to replicate this (i.e. find an example) this morning, so maybe this is just complexity that doesn't need to exist. But I'm reluctant to change it too drastically just yet,

Reply:
So, we have two tests where `self.stat.name != self._iterfit.stat.name` for the `Fit` class, and these are both in `sherpa/som/tests/test_sim.py`. Going through 'git blame` to make sure it wasn't a re-factoring gone wrong, we can see that in the original version 

https://github.com/sherpa/sherpa/blame/111768eafea620bc8dfc8f5a1827b66bbf9f6dad/sherpa/sim/tests/test_sim.py

we have

- `<object>.fit = Fit(data, model, Chi2DataVar(), ...)` and then there is a vall to `<object>.fit.fit()`
- in both `test_mh.py` and test_metropolisMH.py` we then say `<object>.fit.stat = Cash() or CStat()`, which then decoples the `Fit` class

So, we could think about "closing" this loop-hole - e.g. try to avoid allowing the stat to be different by some means, or we can live with the current behaviour and perhaps tweak the documentation pointed out here (although I do agree with @hamogu that this is surprising behaviour).

I do not think that the behavior seen in `test_mh.py` is easily hit from the UI layer - thanks to the way we access the underlying fit object we pretty much always re-create it - so it's only likely an issue for the direct users of the `Fit` class.

Reply:
I created #2063 to track the underlying issue, but we still need to decide what to do here.

Reply:
Thinknig about #2063 made me decide just to "do it". I added (well, added to existing) tests that shows off the old behaviour, so we can see the new behaviour.

The tests are a bit annoying because we don't a-priori know the correct answer, since

a) we are testing an optimizer [in this case a MCMC chain] so we can say "this looks okay" but not "this is definitely correct

b) we don't actually know how the test data was created, which makes it harder to verify "this looks okay" [but we cna see that the parameter values are not changed significantly after fixing up the issue]